Do tree construction in two phases

nox opened the pull request:
First phase builds the block-level boxes on the root element,
then we traverse the resulting intermediate boxes to handle
nested block-level boxes and text shaping.

nox committed:
Do tree construction in two phases

First phase builds the block-level boxes on the root element,
then we traverse the resulting intermediate boxes to handle
nested block-level boxes and text shaping.

nox committed:
Add Debug implementations

SimonSapin commented on victor/src/layout/boxes.rs:
```diff
@@ -33,21 +37,37 @@ pub(super) enum BlockLevel {
 }
+
+#[derive(Debug)]
+pub(super) struct InlineLevelBox {
```
Naming nit: this should be `InlineBox`, or just `Inline` since we’re in the `boxes` module. (All inline boxes are inline-level boxes, but some inline-level boxes are not inline boxes. They can also be text boxes or atomic inline-level boxes.)

nox replied:
It wasn't clear to me that those things actually described boxes without the suffix, so I'll keep it if you don't mind it.

SimonSapin commented on victor/src/layout/boxes/generation.rs:
```diff
@@ -202,3 +412,9 @@ impl<T> Take for Vec<T> {
     }
 }
+
+impl Take for IntermediateInlineFormattingContext {
```
Nit: maybe a generic impl for `T: Default`?

nox committed:
Use rayon during tree construction

nox committed:
Implement Take for T where T: Default

nox committed:
Rename InlineLevelBox to InlineBox

nox committed:
Rename BlockLevel to BlockLevelBox

nox committed:
Rename InlineLevel::Box to InlineLevelBox::InlineBox

nox committed:
Rename IntermediateBlockContainer::Block to BlockBox

nox committed:
Rename Text to TextRun

SimonSapin commented on victor/src/layout/boxes/generation.rs:
```diff
@@ -31,174 +27,433 @@ struct Context<'a> {
+                    (
+                        self.parent_style
+                            .as_ref()
+                            .expect("text node in a document always has a parent"),
```
It looks like you wrote calls to `expect` with a message that should be read as “We expected that […], but that wasn’t the case here.” However the implementation of `expect` contains literally `panic!("{}", msg)` without such context, so panic messages will appear to say the opposite of what you intended. This line should probably be something like `.expect("found a text node without a parent")`.

SimonSapin commented on victor/src/layout/boxes/generation.rs:
```diff
@@ -31,174 +27,433 @@ struct Context<'a> {
+            self.end_ongoing_inline_level_box();
+
+            descendant_node =
+                &self.context.document[descendant_node.parent.expect("descendant has a parent")];
```
Same as above re `expect`.

SimonSapin commented on victor/src/layout/boxes/generation.rs:
```diff
@@ -31,174 +27,433 @@ struct Context<'a> {
+        let parent_style = self.parent_style.as_ref();
+        let anonymous_style = self.anonymous_style.get_or_insert_with(|| {
+            ComputedValues::anonymous_inheriting_from(
+                parent_style.expect("anonymous style is never None when parent style is"),
```
Same as above re `expect`.

SimonSapin commented on victor/src/layout/boxes/generation.rs:
```diff
@@ -31,174 +27,433 @@ struct Context<'a> {
+        let mut last_ongoing_inline_level_box = self
+            .ongoing_inline_level_box_stack
+            .pop()
+            .expect("there is an ongoing inline level box");
```
Same as above re `expect`.

SimonSapin commented on victor/src/layout/boxes/generation.rs:
```diff
@@ -31,174 +27,433 @@ struct Context<'a> {
-impl<Extra: Default + PushBlock> Builder<Extra> {
-    fn new(style: Arc<ComputedValues>) -> Self {
+enum IntermediateBlockContainer {
+    BlockBox(dom::NodeId),
```
I feel that something around here is either misnamed or not structured correctly, but I’m not sure what. When it does not establish an inline formatting context, a block container contains *multiple* block-level boxes, not just one block.

By the way, what is the role of this type v.s. `IntermediateBlockFormattingContext`?

SimonSapin replied:
Ok I think I figured it out. The one `NodeId` here refers not to the child block-level boxes, but to the *parent* DOM element that contains elements that will generate them. So I’d prefer this variant to be:

```rust
    BlockLevelBoxes { generated_by_children_of: dom::NodeId }
```

… or something like that.

----

> what is the role of this type v.s. `IntermediateBlockFormattingContext`?

Ok, so.

At first I thought there was only a not-quite-appropriate name for this Rust type. Then I thought I had found a behavior bug where this PR would create a block formatting context even for “ordinary” blocks that should not create a BFC but instead participate in their parent’s. *Then* I noticed the lone `.0` line that unwraps the `BlockFormattingContext(foo)` Rust value that was just created.

So… here is a series of changes that I hope would make this less misleading. What do you think?

* Rename `IntermediateBlockContainer` → `DeferredNestedBlockContainer` (or is this over-doing it? `DeferredBlockContainer` might work)
* Rename `IntermediateBlockFormattingContext` → `IntermediateBlockContainer`, and change the signature of `finish` accordingly
* Rename `IntermediateBlockFormattingContextBuilder` → `IntermediateBlockContainerBuilder`
* Create the one and only (for now) `BlockFormattingContext(foo)` value in `Document::box_tree`

SimonSapin commented on victor/src/layout/boxes/generation.rs:
```diff
@@ -31,174 +27,433 @@ struct Context<'a> {
+            inline_level_boxes: self.inline_level_boxes,
+            text_runs: self
+                .text_runs
+                .into_par_iter()
```
This is fine for now, but we’ll probably have enough parallelism between separate paragraphs (blocks) that we don’t need to parallelize shaping within a paragraph (one inline formatting context). In fact doing complex shaping correctly (including across inline element boundaries) might require us to sequentialize. Though separating what this module calls `build` and `finish` phases might still be useful so that shaping can have the context of the entire paragraph, so this isn’t wasted effort ;)

(But shaping is for [Skribo](https://raphlinus.github.io/rust/skribo/text/2019/02/27/text-layout-kickoff.html) to solve, naive shaping will do for CSS layout prototyping.)

SimonSapin reviewed (commented):
Sorry I took a while to read through this. It looks very well thought out, congrats! The names feel verbose at first but that does help understanding.

I realize this is asking of you a higher standard than I’ve held myself to in this repository so far, but could you add a code comment in `boxes/generation.rs` that describes on a high level how this module achieves what it does? (The “what” can be a link or two to relevant sections of CSS 2.) In particular, the tree traversal without recursion feels less easy to follow. Can you say (in that code comment) why recursion wouldn’t work here?

nox committed:
Change a few .expect messages

nox committed:
Rename some stuff

IntermediateBlockContainer → DeferredNestedBlockContainer
IntermediateBlockFormattingContext → IntermediateBlockContainer
IntermediateBlockFormattingContextBuilder → IntermediateBlockContainerBuilder

nox committed:
Avoid a call to <Arc<ComputedValues>>::clone

Just because I can.

SimonSapin commented on victor/src/layout/boxes/generation.rs:
```diff
@@ -399,18 +400,31 @@ impl IntermediateBlockLevelBox {
+        self,
+        context: &Context,
+        style: Arc<ComputedValues>,
+    ) -> (BlockContainer, Arc<ComputedValues>) {
```
It looks like this would be simpler if `IntermediateBlockContainerBuilder` held a `&Arc<ComputedStyle>` instead?

nox committed:
Store a reference to parent style in IntermediateBlockContainerBuilder

nox committed:
Improve anonymous style handling

The root most often doesn't have anonymous inline boxes to begin with,
so delaying the initialisation of the anonymous styles for the root
is neat.

nox committed:
Add some documentation
